web/src/search_pill.ts

```diff
@@ -89,45 +99,86 @@ function append_user_pill(
 
+export function is_valid_pill(term: NarrowTerm): boolean {
+    switch (term.operator) {
+        case "has":
+            return ["image", "link", "attachment", "reaction"].includes(term.operand);
+        case "is":
+            return [
+                "dm",
+                "starred",
+                "mentioned",
+                "alerted",
+                "unread",
+                "resolved",
+                "followed",
+            ].includes(term.operand);
+        case "in":
+            return ["home", "all"].includes(term.operand);
+        case "id":
+            return message_store.get(Number.parseInt(term.operand, 10)) !== undefined;
+        case "channel":
+            return stream_data.get_sub(term.operand) !== undefined;
+        case "sender":
+        case "dm":
+        case "dm-including":
+            return term.operand
+                .split(",")
+                .every((email) => people.get_by_email(email) !== undefined);
+        default:
+            return true;
+    }
+}
+
 export function set_search_bar_contents(
     search_terms: NarrowTerm[],
     pill_widget: SearchPillWidget,
-    set_search_bar_text?: (text: string) => void,
+    set_search_bar_text: (text: string) => void,
 ): void {
     pill_widget.clear();
     let partial_pill = "";
+    const invalid_inputs = [];
+
     for (const term of search_terms) {
-        if (user_pill_operators.has(term.operator) && term.operand !== "") {
-            const user_emails = term.operand.split(",");
-            const users = user_emails.map((email) => {
-                const user = people.get_by_email(email);
-                assert(user !== undefined);
-                return user;
-            });
-            append_user_pill(users, pill_widget, term.operator, term.negated ?? false);
-            continue;
-        }
         const input = Filter.unparse([term]);
+
         // If the last term looks something like `dm:`, we
         // don't want to make it a pill, since it isn't isn't
         // a complete search term yet.
         // Instead, we keep the partial pill to the end of the
         // search box as text input, which will update the
         // typeahead to show operand suggestions.
-        if (
-            set_search_bar_text !== undefined &&
-            input.at(-1) === ":" &&
-            term.operand === "" &&
-            term === search_terms.at(-1)
-        ) {
+        if (input.at(-1) === ":" && term.operand === "" && term === search_terms.at(-1)) {
             partial_pill = input;
             continue;
         }
+
+        if (!is_valid_pill(term)) {
+            invalid_inputs.push(input);
+            continue;
+        }
+
+        if (user_pill_operators.has(term.operator) && term.operand !== "") {
+            const users = term.operand.split(",").map((email) => {
+                const user = people.get_by_email(email);
+                assert(user !== undefined);
```

Comment:
Can you add a string on this assert statement for why we believe it to always be valid? I think the reason is we're reading out of an already-completed pill?  `assert(test, "Email should be for a valid user for already-completed pill")` or something.

I think that it might be possible to have that fail if a user's email address changed and we didn't live-update the search query? This should be fixed when we change the pills to store `user_id` rather than email, which I hope to be a near-term follow-up to this PR.

Reply:
It's because `is_valid_pill` checks this directly above.

Reply:
I'll add a code comment, but I'm not sure if we need a string? I think we could even switch this to a `!` instead of an assert -- I'll do that for now.